File: Day47/Part4/chrono_scheduler.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>

using namespace std;

struct Attack {
    long long time;
    long long damage;

    bool operator<(const Attack& other) const {
        return time < other.time;
    }
};

long long solveChronoScheduler(int n, long long d, long long c, vector<Attack>& attacks) {
    sort(attacks.begin(), attacks.end());

    vector<long long> times(n);
    vector<long long> damages(n);
    for (int i = 0; i < n; ++i) {
        times[i] = attacks[i].time;
        damages[i] = attacks[i].damage;
    }

    vector<long long> prefix_damage(n + 1, 0);
    for (int i = 0; i < n; ++i) {
        prefix_damage[i + 1] = prefix_damage[i] + damages[i];
    }

    auto get_range_damage = [&](int start_idx, int end_idx) {
        if (start_idx >= end_idx) return 0LL;
        return prefix_damage[end_idx] - prefix_damage[start_idx];
    };

    vector<long long> dp(n + 1, 0);

    for (int i = n - 1; i >= 0; --i) {
        // cout << "Processing " << i << endl;
        // Option 1: Take the hit
        long long take_hit = damages[i] + dp[i + 1];

        // Option 2: Use Shield
        long long start_time = times[i];
        long long end_active = start_time + d;
        long long end_cooldown = end_active + c;

        // Find first attack > end_active (upper_bound)
        auto it_shield = upper_bound(times.begin(), times.end(), end_active);
        int idx_after_shield = distance(times.begin(), it_shield);

        // Find first attack >= end_cooldown (lower_bound)
        auto it_cooldown = lower_bound(times.begin(), times.end(), end_cooldown);
        int idx_after_cooldown = distance(times.begin(), it_cooldown);

        long long forced_damage = get_range_damage(idx_after_shield, idx_after_cooldown);
        long long use_shield = forced_damage + dp[idx_after_cooldown];

        dp[i] = min(take_hit, use_shield);
    }

    return dp[0];
}

int main() {
    int N = 5;
    long long D = 3;
    long long C = 2;
    vector<Attack> Attacks = {
        {1, 10},
        {2, 20},
        {4, 10},
        {6, 50},
        {8, 5}
    };

    long long result = solveChronoScheduler(N, D, C, Attacks);
    cout << "Minimum Damage: " << result << endl;

    return 0;
}

File: learnLeetCode/Day47/Part4/chrono_scheduler.py
import bisect

def solve_chrono_scheduler(n, d, c, attacks):
    # attacks: list of [time, damage]
    # Sort just in case, though problem says sorted
    attacks.sort()
    
    times = [a[0] for a in attacks]
    damages = [a[1] for a in attacks]
    
    # Prefix sum for damages
    # prefix_damage[i] = sum(damages[0]...damages[i-1])
    prefix_damage = [0] * (n + 1)
    for i in range(n):
        prefix_damage[i+1] = prefix_damage[i] + damages[i]
        
    def get_range_damage(start_idx, end_idx):
        # Sum of damages from index start_idx to end_idx - 1
        if start_idx >= end_idx:
            return 0
        return prefix_damage[end_idx] - prefix_damage[start_idx]

    # dp[i] = min damage from i to n-1
    dp = [0] * (n + 1)
    
    for i in range(n - 1, -1, -1):
        # Option 1: Take the hit
        take_hit = damages[i] + dp[i+1]
        
        # Option 2: Use Shield
        start_time = times[i]
        end_active = start_time + d
        end_cooldown = end_active + c
        
        # Find first attack that is NOT blocked (time > end_active)
        # bisect_right returns insertion point after all x <= val.
        # We want first x > val. So bisect_right is correct.
        idx_after_shield = bisect.bisect_right(times, end_active)
        
        # Find first attack that is available after cooldown (time >= end_cooldown)
        # bisect_left returns insertion point before all x >= val.
        idx_after_cooldown = bisect.bisect_left(times, end_cooldown)
        
        # Damage taken during cooldown period (unavoidable if we shield now)
        # Attacks from idx_after_shield to idx_after_cooldown - 1
        forced_damage = get_range_damage(idx_after_shield, idx_after_cooldown)
        
        use_shield = forced_damage + dp[idx_after_cooldown]
        
        dp[i] = min(take_hit, use_shield)
        
    return dp[0]

if __name__ == "__main__":
    N = 5
    D = 3
    C = 2
    Attacks = [
        [1, 10],
        [2, 20],
        [4, 10],
        [6, 50],
        [8, 5]
    ]
    
    result = solve_chrono_scheduler(N, D, C, Attacks)
    print(f"Minimum Damage: {result}")
